Fix repainting for animated carets

mwyrzykowski opened the pull request:
#### 6adf3d29d4b6bd80d164bdc6a4e15edb36bb0b9c
<pre>
Fix repainting for animated carets
<a href="https://bugs.webkit.org/show_bug.cgi?id=256619">https://bugs.webkit.org/show_bug.cgi?id=256619</a>
&lt;radar://109174040&gt;

Reviewed by Simon Fraser.

Simply using *partialRepaintRect is wrong. We should still be calling
computeRectForRepaint(...) and adjusting the rect as needed.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::computeOutsetsFromInnerOuterRect):
(WebCore::repaintCaretForLocalRect):
Pass margins of the expanded rectangle.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::issueRepaint const):
(WebCore::RenderObject::repaintRectangle const):
* Source/WebCore/rendering/RenderObject.h:
Expand the rectangle after calling computeRectForRepaint

Canonical link: <a href="https://commits.webkit.org/264023@main">https://commits.webkit.org/264023@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6beaddfda149267bf070a8163a99b56e3bc3ab8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6455 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8034 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6750 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6451 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6974 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6614 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/9644 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6568 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6618 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5849 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8113 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4083 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5829 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13679 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6000 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5908 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8225 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6391 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5797 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1522 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9956 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6168 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -545,6 +545,7 @@ void Recorder::drawControlPart(ControlPart& part, const FloatRoundedRect& border
 
 void Recorder::resetClip()
 {
+    currentState().clipBounds = FloatRect::infiniteRect();
```
I wonder if this is testable in any way?

mwyrzykowski replied:
I think I can add a test to `DisplayListRecorderTests.cpp`. I will look at that as the review is on-going

mwyrzykowski replied:
I added a test over in https://github.com/WebKit/WebKit/pull/13734

I will remove these changes as the two changes are not really dependent on each other and there were already separate bugs tracking them

smfr commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -2422,6 +2422,18 @@ void RenderBlock::computeBlockPreferredLogicalWidths(LayoutUnit& minLogicalWidth
+    RenderBlock* caretPainter = frame().selection().caretRendererWithoutUpdatingLayout();
+    bool isContentEditable = frame().selection().selection().hasEditableStyle();
+
+    if (result && caretPainter == this && (isContentEditable || settings().caretBrowsingEnabled()))
```
This is going back towards the approach we were trying to avoid with earlier changes.

RenderBlock shouldn't have to know about caretPainter behavior here. We need to pass some state down with the repaint that originates from the caret painter.

mwyrzykowski replied:
Isn't that what we already do on FrameSelection.cpp:1767 with the call to `repaintRectangle`? That bypasses the call to `computeRectForRepaint` and returns the adjusted rectangle as we do here, but maybe there is an additional clip occurring in `RenderObject::repaintUsingContainer`?

mwyrzykowski replied:
Discussed offline and I will take a closer look to see why this occurs on quip and in text input fields

mwyrzykowski commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -1010,9 +1010,12 @@ void RenderObject::issueRepaint(std::optional<LayoutRect> partialRepaintRect, Cl
 
     LayoutRect repaintRect;
-    if (partialRepaintRect)
-        repaintRect = clipRepaintToContainer == ClipRepaintToContainer::No ? *partialRepaintRect : computeRectForRepaint(*partialRepaintRect, repaintContainer.renderer);
```
@smfr thanks for pushing on this, much appreciated. The root cause of the bug was simply using `*partialRepaintRect` which is wrong.

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1757,14 +1757,25 @@ IntRect FrameSelection::absoluteCaretBounds(bool* insideFixed)
     return m_absCaretBounds;
 }
 
+static LayoutBoxExtent computeMarginsFromInnerOuterRect(const LayoutRect& innerRect, const LayoutRect& outerRect)
```
"margins" is an overloaded term in CSS/layout. Maybe use "outset"

smfr commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1757,14 +1757,25 @@ IntRect FrameSelection::absoluteCaretBounds(bool* insideFixed)
+static LayoutBoxExtent computeMarginsFromInnerOuterRect(const LayoutRect& innerRect, const LayoutRect& outerRect)
+{
+    LayoutBoxExtent result;
+    result.setLeft(innerRect.x() - outerRect.x());
```
You should probably clamp these to >= 0.

smfr commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -661,10 +661,9 @@ class RenderObject : public CachedImageClient {
     enum class ClipRepaintToLayer : bool { No, Yes };
     enum class ForceRepaint : bool { No, Yes };
-    void repaintRectangle(const LayoutRect&, ClipRepaintToLayer, ForceRepaint, ClipRepaintToContainer) const;
+    void repaintRectangle(const LayoutRect&, ClipRepaintToLayer, ForceRepaint, std::optional<LayoutBoxExtent> = std::nullopt) const;
```
The last argument should be named. I would call it "additionalRepaintOutsets"

smfr commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -832,7 +831,7 @@ class RenderObject : public CachedImageClient {
     bool isSetNeedsLayoutForbidden() const;
 
-    void issueRepaint(std::optional<LayoutRect> partialRepaintRect = std::nullopt, ClipRepaintToLayer = ClipRepaintToLayer::No, ForceRepaint = ForceRepaint::No, ClipRepaintToContainer = ClipRepaintToContainer::Yes) const;
+    void issueRepaint(std::optional<LayoutRect> partialRepaintRect = std::nullopt, ClipRepaintToLayer = ClipRepaintToLayer::No, ForceRepaint = ForceRepaint::No, std::optional<LayoutBoxExtent> = std::nullopt) const;
```
Ditto.

mwyrzykowski committed:
Fix repainting for animated carets
https://bugs.webkit.org/show_bug.cgi?id=256619
<radar://109174040>

Reviewed by Simon Fraser.

Simply using *partialRepaintRect is wrong. We should still be calling
computeRectForRepaint(...) and adjusting the rect as needed.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::computeOutsetsFromInnerOuterRect):
(WebCore::repaintCaretForLocalRect):
Pass margins of the expanded rectangle.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::issueRepaint const):
(WebCore::RenderObject::repaintRectangle const):
* Source/WebCore/rendering/RenderObject.h:
Expand the rectangle after calling computeRectForRepaint

Canonical link: https://commits.webkit.org/264023@main
